### PR TITLE
Fix issue when client disconnects midway through a stream

### PR DESCRIPTION
**Pull request by 0xTim**
Fixes an issue when a client disconnects mid way through streaming a request in a Swift concurrency context. In certain cases this would trigger a de-init off the event loop, leading to a crash.

This fixes the issue by using a lock instead of a loop bound wrapper

### CODE REVIEW

File: Sources/Vapor/Request/Request+BodyStream.swift [hunk 1]
```diff
@@ -16,13 +16,13 @@ extension Request {
             var buffer: [(BodyStreamResult, EventLoopPromise<Void>?)]
         }
 
-        private let isClosed: NIOLoopBoundBox<Bool>
+        private let isClosed: NIOLockedValueBox<Bool>
```
Comment (MahdiBM) [inline Sources/Vapor/Request/Request+BodyStream.swift:19]: Shouldn't an atomic be better?

Comment (gwynne) [reply Sources/Vapor/Request/Request+BodyStream.swift:19]: No.

**Comment by gwynne** [comment]
@0xTim I'm not delighted with this solution; I'd much rather understand the underlying reason that an early disconnect causes a wrong-thread reinit.

File: Sources/Vapor/Request/Request+BodyStream.swift (hunk 1 above)
Comment (MahdiBM) [reply Sources/Vapor/Request/Request+BodyStream.swift:19]: Why? 🤔

**Review by gwynne — APPROVED** [review]
Reluctantly.

**Pull request merged**